src/npc_attack.cpp

```diff
@@ -0,0 +1,652 @@
+#include "npc_attack.h"
+
+#include "character.h"
+#include "flag.h"
+#include "game.h"
+#include "item.h"
+#include "magic.h"
+#include "magic_spell_effect_helpers.h"
+#include "map.h"
+#include "messages.h"
+#include "npc.h"
+#include "point.h"
+#include "projectile.h"
+#include "ranged.h"
+
+namespace npc_attack_constants
+{
+const std::map<Creature::Attitude, int> attitude_multiplier = {
+    { Creature::Attitude::FRIENDLY, -10 },
+    { Creature::Attitude::HOSTILE, 3 },
+    { Creature::Attitude::NEUTRAL, -1 }
+};
+// if you are attacking your target, multiply potential by this number
+const int target_modifier = 3;
+// if you kill the creature, multiply the potential by this number
+const int kill_modifier = 2;
+// the amount of penalty if the npc has to change what it's wielding
+// update this number and comment when that is no longer a flat -15 moves
+const int base_time_penalty = 3;
+// we want this out of our hands, pronto.
+// give a large buff to the attack value so it prioritizes this
+const int base_throw_now = 10'000;
+} // npc_attack_constants
+
+// TODO: make a better, more generic "check if this projectile is blocked" function
+// TODO: put this in a namespace for reuse
+static bool has_obstruction( const tripoint &from, const tripoint &to )
+{
+    std::vector<tripoint> line = line_to( from, to );
+    // @to is what we want to hit. we don't need to check for obstruction there.
+    line.pop_back();
+    const map &here = get_map();
+    for( const tripoint &line_point : line ) {
+        if( here.impassable( line_point ) ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool npc_attack_effectiveness::operator>( const npc_attack_effectiveness &rhs ) const
```

Comment:
can we call this structure `npc_attack_rating` and save 7 characters per use? Not a big deal, just a thought.